### DISCUSSION TITLE
Hydrogen as JACK timebase slave - can't play drumkit freely

### DISCUSSION THREAD

**Discussion by ttshaw1** [hydrogen-music/hydrogen#2029] [Q&A] [answered]
I'm noticing some weirdness when I try to use Ardour as the JACK timebase master. When JACK is not connected, is the master, or is "out of sync", Hydrogen works independently. I can play a song or loop a pattern and use my MIDI synth to add drums on top of it, with or without recording them. 

But when Hydrogen is slaved to Ardour and they're "in sync", Hydrogen's MIDI-In LED lights up when I press a synth key, but I don't get any audio out. It also sounds like the drum track is doubled. It's certainly much louder. Sample attached; in the first part, Hydrogen is in sync. In the second part, it's out of sync.

https://github.com/user-attachments/assets/b25ccc19-1c06-4a0c-95aa-a1f2b7b72353

To clarify what I mean about in vs. out of sync, I can have Ardour playing and dictating the time with Hydrogen in song mode, but if I switch to Pattern mode, Play, then back to Song and Play, they're out of sync and hydrogen doesn't care what's going on in Ardour. There are other patterns of messing with Pattern/Song mode and Play/Stop that will desync them.

I don't want to file this as an Issue because I might be misunderstanding something, being new to this software. Is there a way to have Hydrogen be the slave in song mode and still be able to throw in additional drums on top of what's being played? Is the doubled drums normal? Would it be better to use Hydrogen as the master as suggested [here](https://discourse.ardour.org/t/connect-and-synchronize-hydrogen-and-ardour/102415)? That feels unnatural to me because I see Ardour as the center of the setup - everything should eventually feed into Ardour.

**Comment by theGreatWhiteShark** [comment]
> Is there a way to have Hydrogen be the slave in song mode and still be able to throw in additional drums on top of what's being played?

Ideally this should work out of the box. But JACK Timebase is, unfortunately, not a good standard. It has massive loop holes, like not being able to handle tempo changes properly (be aware of that) and due to a missing reference implementation or test suite there are various sorts of little bugs in a lot of client applications. Those in Hydrogen I tried to tackle for the upcoming release in #1977.

Could you check whether you have to same issues using [this](https://ci.appveyor.com/api/buildjobs/25125m5h433pvg3f/artifacts/Hydrogen-1.2.3-274-g5dd5e89ff-x86_64.AppImage) version? (It corresponds to the latest commit of branch `releases/1.2` in case you prefer compiling it yourself)

> Would it be better to use Hydrogen as the master as suggested [here](https://discourse.ardour.org/t/connect-and-synchronize-hydrogen-and-ardour/102415)?

No. Both modes should work. If not its a bug and we will (try to) fix it. That said, Master mode tells Hydrogen not to listen to external time signals and to just send them. There is a lot less things that could go wrong in there. But again, non-Master should work too.

> Un-check “Ardour is JACK Time Master”.

This is one of the little things that make JACK Timebase really hard to support. When checked (at least at the time I read their code for debugging) Ardour won't be Timebase Master at all times. This is not how this protocol works. It only makes Ardour discard external time signals (and maybe tells it to periodically attempt to register as master again).

> That feels unnatural to me because I see Ardour as the center of the setup - everything should eventually feed into Ardour.

Yes. That's how I would use it too.

**Reply by ttshaw1** [reply]
Thank you very much for sharing. I tried out the AppImage you linked and it solve the two problems I talked about initially; with Ardour dictating the timing, I can add drums over what's in the pattern as the timeline plays, and the audio sounds normal.

I did notice v1.2.3 starts up in `JACK Timebase mode = Listener`, while the AppImage version started in `JACK Timebase mode = Off`. And more significantly, it seems like when v1.2.3 is in sync with Ardour, it's much more closely slaved to Ardour's timing. v1.2.3 will exactly mirror Ardour's measure number position and play/stop status. The AppImage version will follow Ardour's BPM (which is cool, nice fix) and on the occasions where it needs to jump to a position in the timeline, will jump to Ardour's position. But it doesn't respect Ardour's play/stop status or follow timeline jumps I make in Ardour. 

Possibly this is due to the new Hydrogen version starting in `Timebase mode = Off`; I've been testing this by either switching it to Timebase mode = Master then restarting Ardour so Ardour will grab Master status then toggling Hydrogen to Slave, or setting Hydrogen to Timebase mode = Slave and restarting Ardour. For all I know that could be insufficient to get the timebase set up properly.

If this behavior is known or expected then I'm happy to close this discussion. But if it's unintended please let me know if there's any other info you want from me.

**Reply by theGreatWhiteShark** [reply]
> And more significantly, it seems like when v1.2.3 is in sync with Ardour, it's much more closely slaved to Ardour's timing.

Hmm. Sounds like a regression. Is this a constant offset, does it change in time or is e.g. tempo dependent?

> I did notice v1.2.3 starts up in JACK Timebase mode = Listener, while the AppImage version started in JACK Timebase mode = Off

So, when starting playback in Ardour the "J.MASTER" button in v1.2.3 is turning blue but in the AppImage it does not?

> The AppImage [...] doesn't respect Ardour's play/stop status or follow timeline jumps I make in Ardour.

Hmm. This sounds like Ardour is not sending the proper JACK commands. But I would need to see some logs to verify that.

> If this behavior is known or expected then I'm happy to close this discussion. But if it's unintended please let me know if there's any other info you want from me.

No, that is not expected and I'm very glad you are bringing this up prior to the 1.2.4 release :).

Could you tell which OS and Ardour version you are using? I mainly developed the JACK Timebase code against Muse4. Ardour never worked for me. I tried major version 1, 2, and 3 some 10 years ago but it was way too unstable and crashed all the time. For the new Timebase stuff I also intended to test against Ardour, compiled the latest version but crashes during startup before even getting to a splash screen. Where do you retrieved your Ardour version?

Ideally I would get Ardour (maybe using your OS) running and could be able to reproduce your issues locally. If not, we would need to resort to logging on your end (basically replacing the `0` in [this line](https://github.com/hydrogen-music/hydrogen/blob/releases/1.2/src/core/IO/JackAudioDriver.cpp#L55) by `1`). I would make you a patched AppImage and it would be great if you could help debugging 😉.

**Reply by ttshaw1** [reply]
> Hmm. Sounds like a regression. Is this a constant offset, does it change in time or is e.g. tempo dependent?

I'm not sure which case you're referring to. In v1.2.3, there's no offset at all, assuming they're in sync. See the webm; they're out of sync until 16 seconds, when a Rewind in Hydrogen gets them in sync. Sorry if the video's choppy. From then on, until I desync them, I expect Hydrogen to be exactly where Ardour is regardless of tempo or about anything else. If the tempo in Hydrogen and Ardour don't match, the individual drum sounds will be faster or slower than usual, but they start where they're supposed to and the timeline is where it should be. 
[Screencast_20240915_225223.webm](https://github.com/user-attachments/assets/e7c982af-0ea3-4288-a007-e872d7669799)

In the AppImage version, I can get Hydrogen's timeline to jump to where Ardour's currently is, and the tempo always matches Ardour's. So if Ardour is playing, and I jump in Hydrogen to Ardour's current position, now they'll be together in time until I change something. But if I pause or jump around the timeline in Ardour, Hydrogen doesn't care and will keep right on playing as normal. 

> So, when starting playback in Ardour the "J.MASTER" button in v1.2.3 is turning blue but in the AppImage it does not?

Right. 
![image](https://github.com/user-attachments/assets/fd733b15-62cb-4212-b437-99a065d4c434)
![image](https://github.com/user-attachments/assets/bb23cde8-955c-408c-9671-22c5e9062746)
The red outline around the CPU bar does go away after a second or two.

> Could you tell which OS and Ardour version you are using? I mainly developed the JACK Timebase code against Muse4. 

I'm running Arch with kernel Linux 6.10.10-zen1-1-zen x86_64. I use the Ardour version in the official Arch repositories, currently v8.6.2. 
 
> Ideally I would get Ardour (maybe using your OS) running and could be able to reproduce your issues locally. If not, we would need to resort to logging on your end (basically replacing the `0` in [this line](https://github.com/hydrogen-music/hydrogen/blob/releases/1.2/src/core/IO/JackAudioDriver.cpp#L55) by `1`). I would make you a patched AppImage and it would be great if you could help debugging 😉.

I've found Ardour to be pretty stable on my setup. I imagine it's come quite a long way since v3. But if it couldn't get past the splash screen maybe it's very dependent on the setup. If you want to try to reproduce my setup, here's a list of the packages I've installed. Dependencies aren't shown but Arch's package manager is smart enough to grab them if you install the list.
[packages.txt](https://github.com/user-attachments/files/17008561/packages.txt)

Certainly you could get more information and faster versus having me test with a logging version, but it sounds like a lot of effort to set up Arch for this. If you want to give me an AppImage with logging enabled I'd be happy to test it. 

I also realized I haven't messed around with different time signatures. I'm not sure what impact that'll have.

Are you aiming to have the code for v1.2.4 frozen in the next few days? I can put in a half hour here and there on weeknights, but I may not be able to really dig deep until the weekend. 

Thanks for the support on this. I just did my first legitimate songwriting session today and it blew me away how important the drums I sequenced in Hydrogen were. This is fantastic software.

**Reply by theGreatWhiteShark** [reply]
> The red outline around the CPU bar does go away after a second or two.

This indicates a so-called XRun event. Meaning the audio engine was not able to process its audio in time and JACK discards the content of Hydrogen's audio buffer for that processing cycle.

> I've found Ardour to be pretty stable on my setup.

NIce. I by no means want to discourage its usage.

> If you want to try to reproduce my setup, here's a list of the packages I've installed. Dependencies aren't shown but Arch's package manager is smart enough to grab them if you install the list.
[packages.txt](https://github.com/user-attachments/files/17008561/packages.txt)

Thank's a lot. I'll do in a couple of days.

> Are you aiming to have the code for v1.2.4 frozen in the next few days?

No. I mean, the sooner the better. But we are in not in a hurry and quality and stability is a lot more important for us. In addition, I'm currently working on another issue for 1.2.4 as well.

> I can put in a half hour here and there on weeknights, but I may not be able to really dig deep until the weekend.

Similar for me. But I have less time during weekends :D

**Reply by theGreatWhiteShark** [reply]
> > If you want to try to reproduce my setup, here's a list of the packages I've installed. Dependencies aren't shown but Arch's package manager is smart enough to grab them if you install the list.
[packages.txt](https://github.com/user-attachments/files/17008561/packages.txt)

> Thank's a lot. I'll do in a couple of days.

What's up with Arch Linux? Haven't used it in a while. They do not feature a graphical installation anymore? One has to write a `wpa_supplicant` configuration by hand and start it as a background process when just having a single shell? Oh boy, even Void Linux has a better UX. 😆 But anyway. I can't get it to run on my second laptop as its WLAN is not properly working with their shipped kernel. I'll try with a different Linux audio distribution which comes with a preinstalled Ardour in a couple of days.

**Reply by ttshaw1** [reply]
I installed something like a couple years ago and it's definitely a slog to set up. I can't remember if there was a GUI at the time but I opted to stick close to the Arch Wiki install guide, which is all command line. From what I can remember, going through the critical steps [here](https://wiki.archlinux.org/title/Installation_guide) was enough to get the computer functional enough that when I installed KDE Plasma, everything else was easy to set in a GUI. But I think I was working from home at the time, so god knows how many hours it took to actually do that. 

If you can get the info you need if I run an AppImage with logging please do send one over. I imagine the total man hours with that approach will be something like 10% of what it will take for you to get a working Arch install up.

**Reply by theGreatWhiteShark** [reply, marked as answer]
Don't worry. We just have to deal with JACK and an another JACK client in here. Other distributions should yield the same results. I just set up a new AV Linux machine with Ardour running.

First thing I noticed is that Ardour does neither send JACK positions nor does it start/stop JACK transport out of the box. In the upper left part there is a button reading `Int.`. Right clicking it allows you to select JACK as "Transport Master".  Close the dialog and click the `Int.` button until it turns green and displays `JACK`. Now Hydrogen should pick up transport immediately.

**Reply by ttshaw1** [reply]
@theGreatWhiteShark sorry I didn't get back to you. I somehow had the misconception that there was something you were looking into and the ball was in your court, and I managed to neglect looking at this thread for ages. Turns out, you're completely correct, Ardour was set to use a MIDI Timebase. With Ardour set to JACK Transport Master and Hydrogen set as JACK Timebase Controller, things are working smoothly. Moving the playhead in either program moves it in the other, and play/pause in one program also plays/pauses in the other. Thank you so much for your help. I'm including screenshots in case anyone else runs into this in the future.

![image](https://github.com/user-attachments/assets/d08b3ef2-6e51-450b-9f90-890ce37f88b8)
![image](https://github.com/user-attachments/assets/1573b8a0-810a-4372-a630-4692f9ba9730)
